### ISSUE TITLE
How to handle non-sequential request ID messages in a session?

### ISSUE THREAD

**Issue by SunFulong** [wamp-proto/wamp-proto#436]
Hello,

I'm implementing my WAMP router and client library, the spec defines "IDs in the session scope MUST be incremented by 1 beginning with 1 (for each direction)", but not describe what should peer do if received non-sequential request ID.

So, what should I do if received an unexpected request ID from peer?

Thanks!

**Comment by KSDaemon** [comment]
Hi! 

Well, AFAIK, nothing bad should happen. In some implementations, I'm aware of, that is not checked, so there is no problem.

We might need to clarify this in spec.

**Comment by SunFulong** [comment]
How about defining a strict mode that will ABORT the session immediately if the request ID is not sequential?

**Comment by ecorm** [comment]
Perhaps it should be "IDs in the session scope MUST be monotonic and SHOULD be incremented by 1 beginning with 1".

There may be use cases where a client implementation picks the request ID for an outgoing CANCEL message, but client decides to discard that message from its queue because the RPC completes before the CANCEL message is sent.

**Comment by oberstet** [comment]
> but not describe what should peer do if received non-sequential request ID.

personally, and in general, I'd opt for "protocol error"

https://wamp-proto.org/wamp_latest_ietf.html#name-protocol-errors

that is harsh, but why continue talking to a broken peer? that's a risk.

in this specific case, I would expect most routers not really check that currently. should we add a UDP transport, this ID sequencing (strictly monotonic without gaps) might become actually useful .. one reason I'm -1 on changing the spec text and loosing it to "monotonic".

**Comment by ecorm** [comment]
> but client decides to discard that message from its queue because the RPC completes before the CANCEL message is sent.

I should have noted that it should be harmless for the client to leave the defunct CANCEL in its transmit queue and allow it to be sent to the dealer. The dealer should just ignore it if the RPC already completed.

**Comment by oberstet** [comment]
> The dealer should just ignore it if the RPC already completed.

Alright, so it's "valid behavior" for a client to send CANCEL to a router at 1k requests/sec with made up request IDs?

**Comment by ecorm** [comment]
@oberstet What I'm saying is that the dealer should be able to handle the race condition where it receives a CANCEL message (due to a timeout on the caller side) after it has already sent the RESULT response. Note that the CANCEL message itself does not have its own request ID. I had raised this with #160, but I don't think it ever got resolved.

Having said that, a dealer could keep track of the request ID of the most recent CALL message, and treat a CANCEL message with a greater request ID as suspicious/erroneous.

I suppose a dealer could also "remember" the request IDs of completed RPCs for a limited time, and not panic if it receives a CANCEL message with one of those remembered request IDs.

**Comment by ecorm** [comment]
> Alright, so it's "valid behavior" for a client to send CANCEL to a router at 1k requests/sec with made up request IDs?

To play the Devil's Advocate: couldn't a client just as easily mount a DoS attack with 1k messages/sec that are valid protocol-wise?

**Comment by KSDaemon** [comment]
> in this specific case, I would expect most routers not really check that currently. should we add a UDP transport, this ID sequencing (strictly monotonic without gaps) might become actually useful .. one reason I'm -1 on changing the spec text and loosing it to "monotonic".

Well. WAMP is an application-level protocol at its best. So I don't think we should introduce any transport level specifics to it.   It will be really hard to take into account all possible edge cases and scenarios. That's why we simply state that message order in the peer queue must be guaranteed (no matter how). 

And yeah, I also think that not routers nor clients do not actually check that sequence numbers and operate only on message types and related structures (existing calls, invocations, etc).

> Having said that, a dealer could keep track of the request ID of the most recent CALL message, and treat a CANCEL message with a greater request ID as suspicious/erroneous.
> I suppose a dealer could also "remember" the request IDs of completed RPCs for a limited time, and not panic if it receives a CANCEL message with one of those remembered request IDs.

Good points! 

And regarding DoS Attacks: Guys, do you really think it is possible to mitigate malicious attacks on the protocol level? Name even one network protocol that successfully does it! So protection is definitely another story. An implementation may take care of that on their level but it is definitely not on the protocol's shoulders itself. And as Emile well noted: hackers may DDoS even with valid requests if they'd like to.

**Comment by SunFulong** [comment]
> And yeah, I also think that not routers nor clients do not actually check that sequence numbers and operate only on message types and related structures (existing calls, invocations, etc).

In my opinion, the rules stated in the specification, all implementations MUST follow. 
Session scoped request IDs are sequential is defined behavior, even existing clients or routers don't check it. But, if they connect to a "strict peer", they may be denied future requests.

**Comment by oberstet** [comment]
> should be able to handle the race condition where it receives a CANCEL message (due to a timeout on the caller side) after it has already sent the RESULT response

yes, you are right, I forgot again: there is a race between CANCEL and either 1) sending RESULT back to the caller or 2) receiving YIELD from the callee. a WAMP transport is ordered, but for each direction (up-/downstream) only, it's not totally ordered. so the race is unavoidable. the only way is for the Router to ignore, not error on such out-raced CANCEL message.

in this case, we're missing some clarification in https://wamp-proto.org/wamp_latest_ietf.html#section-11.4-17 I guess

> Having said that, a dealer could keep track of the request ID of the most recent CALL message, and treat a CANCEL message with a greater request ID as suspicious/erroneous.

`CALL.Request|id` is [sequential at session scope](https://wamp-proto.org/wamp_latest_ietf.html#section-6.2.1-8.1), and so yes, a CANCEL with an ID higher than any previous CALL is definitely bogus.

a `CANCEL.Request|id` for an old ID that never was used for a `CALL` before is also bogus, but costs memory to track

> I suppose a dealer could also "remember" the request IDs of completed RPCs for a limited time, and not panic if it receives a CANCEL message with one of those remembered request IDs.

this is even costlier: store and track how long stored and only then delete. but yes, that's one option.

another one: track how often CANCEL was received per request ID.

which brings me to this Q: is it valid for client to send CANCEL *multiple times* for one previous, unfinished CALL?

until it received an ERROR (with "canceled" as reason) or a RESULT (in case the cancel lost the race)?

> To play the Devil's Advocate: couldn't a client just as easily mount a DoS attack with 1k messages/sec that are valid protocol-wise?

yeah, I see what you mean. Still, there is a difference between "intentionally overloading a service in a formally proper way" or "intentionally overloading a service by misformed workload" doesn't matter in the end.

> And regarding DoS Attacks: Guys, do you really think it is possible to mitigate malicious attacks on the protocol level? Name even one network protocol that successfully does it! 

We should do what we can, follow best practice, but in general, I agree, the mitigation must happen somewhere else.

rgd protocols: eg SIP and CoAP have issues/fixes wrt amplification attacks. here is an older list: https://www.cisa.gov/uscert/ncas/alerts/TA14-017A. QUIC is very strict https://www.ietf.org/archive/id/draft-mattsson-t2trg-amplification-attacks-01.html#name-summary

The ideal attacker protocol allows an attacker to have almost zero cost for making a service impose a high cost onto a 3rd party, and have the service pay for the cost of imposing load on the 3rd party.

> Session scoped request IDs are sequential is defined behavior, even existing clients or routers don't check it now. But, this is an open protocol, if someday, they talk to a "strict peer", their requests may be denied.

Exactly! The protocol should be defined in a strict way even when implementations not yet enforce it fully. But they should of course.

FWIW, I've researched the approaches in design philosophy a bit:

1. "be conservative in what you do, be liberal in what you accept from others": [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle) - the Internet old-school
2. [Maintaining Robust Protocols](https://datatracker.ietf.org/doc/html/draft-iab-protocol-maintenance): the Internet new-school

here is a quote from the latter:

> A flaw can become entrenched as a de facto standard. Any implementation of the protocol is required to replicate the aberrant behavior, or it is not interoperable. This is both a consequence of applying the robustness principle, and a product of a natural reluctance to avoid fatal error conditions. Ensuring interoperability in this environment is often referred to as aiming to be "bug for bug compatible".

**Comment by ecorm** [comment]
> WAMP is an application-level protocol at its best. So I don't think we should introduce any transport level specifics to it.

I tend to agree with this view. If a UDP WAMP transport is ever proposed, it should have its own mechanisms for reliability. I was curious if a protocol already existed for reliable UDP and found this: https://en.wikipedia.org/wiki/Reliable_User_Datagram_Protocol

> which brings me to this Q: is it valid for client to send CANCEL multiple times for one previous, unfinished CALL?
> until it received an ERROR (with "canceled" as reason) or a RESULT (in case the cancel lost the race)?

I can imagine an UI emitting multiple CANCEL messages for the same CALL due to an impatient user, where neither the UI nor the WAMP client keeps track of the already pending cancellation.

**Comment by meejah** [comment]
> do you really think it is possible to mitigate malicious attacks on the protocol level?

Not really [*].

However, protocols _can_ make it way easier for DoS attacks, so it's still important to be careful here. Look at the disaster of the VoIP protocols :) for an example. Anyway, in this particular case I agree that there's easier (protocol-valid) ways to DoS than spamming CANCEL messages.

[*] -- it's possible to include DoS-mitigation features "in" a protocol. For example, the magic-wormhole protocols now have an (experimental, PoC) mechanism to demand "hashcash" (i.e. proof-of-work") from clients. This can _help_ mitigate a DoS, in theory, similar to its original use in email: legitimate clients are happy to wait a few hundred milliseconds but it means you need a lot of CPU to do the bad stuff.

One could even argue that authentication is a form of DoS mitigation: you have to prove you have a particular account to proceed (and then an admin can deny you service entirely if you've been abusive). Of course, that kind of just pushes the problem to "how hard is it to get an account" instead, but ...
